src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+	overloadAndRemove(std::vector<ProtoPalette> const &protoPalettes) {
+	options.verbosePrint("Paginating palettes using \"overload-and-remove\" strategy...\n");
+
+	struct Iota {
+		using value_type = size_t;
+		using difference_type = size_t;
+		using pointer = value_type const *;
+		using reference = value_type const &;
+		using iterator_category = std::input_iterator_tag;
+
+		// Use aggregate init etc.
+		value_type i;
+
+		bool operator!=(Iota const &other) { return i != other.i; }
+		reference operator*() const { return i; }
+		pointer operator->() const { return &i; }
+		Iota operator++() {
+			++i;
+			return *this;
+		}
+		Iota operator++(int) {
+			Iota copy = *this;
+			++i;
+			return copy;
+		}
+	};
+
+	// Begin with all proto-palettes queued up for insertion
+	std::queue queue(std::deque<ProtoPalAttrs>(Iota{0}, Iota{protoPalettes.size()}));
+	// Begin with no pages
+	std::vector<AssignedProtos> assignments{};
+
+	for (; !queue.empty(); queue.pop()) {
+		ProtoPalAttrs const &attrs = queue.front(); // Valid until the `queue.pop()`
+
+		ProtoPalette const &protoPal = protoPalettes[attrs.palIndex];
+		size_t bestPalIndex = assignments.size();
+		// We're looking for a palette where the proto-palette's relative size is less than
+		// its actual size; so only overwrite the "not found" index on meeting that criterion
+		double bestRelSize = protoPal.size();
+
+		for (size_t i = 0; i < assignments.size(); ++i) {
+			// Skip the page if this one is banned from it
+			if (attrs.isBannedFrom(i)) {
+				continue;
+			}
+
+			options.verbosePrint("%zu: Rel size: %f (size = %zu)\n", i,
+			                     assignments[i].relSizeOf(protoPal), protoPal.size());
+			if (assignments[i].relSizeOf(protoPal) < bestRelSize) {
+				bestPalIndex = i;
+			}
+		}
+
+		if (bestPalIndex == assignments.size()) {
+			// Found nowhere to put it, create a new page containing just that one
+			assignments.emplace_back(protoPalettes, std::move(attrs));
+		} else {
+			auto &bestPal = assignments[bestPalIndex];
+			// Add the color to that palette
+			bestPal.assign(std::move(attrs));
+
+			// If this overloads the palette, get it back to normal (if possible)
+			while (bestPal.volume() > 4) {
+				options.verbosePrint("Palette %zu is overloaded! (%zu > 4)\n", bestPalIndex,
+				                     bestPal.volume());
+
+				// Look for a proto-pal minimizing "efficiency" (size / rel_size)
+				auto efficiency = [&bestPal](ProtoPalette const &pal) {
+					return pal.size() / bestPal.relSizeOf(pal);
+				};
+				auto [minEfficiencyIter, maxEfficiencyIter] =
+					std::minmax_element(bestPal.begin(), bestPal.end(),
+				                        [&efficiency, &protoPalettes](ProtoPalAttrs const &lhs,
+				                                                      ProtoPalAttrs const &rhs) {
+											return efficiency(protoPalettes[lhs.palIndex])
+					                               < efficiency(protoPalettes[rhs.palIndex]);
+										});
```

Comment:
Not sure if this is something we'd agree on or even something clang-format could do, but here's how I'd format this:
```suggestion
				auto [minEfficiencyIter, maxEfficiencyIter] =
					std::minmax_element(
						bestPal.begin(),
						bestPal.end(),
				        [&efficiency, &protoPalettes](
				        	ProtoPalAttrs const &lhs,
				            ProtoPalAttrs const &rhs
				        ) {
							return efficiency(protoPalettes[lhs.palIndex])
					    		< efficiency(protoPalettes[rhs.palIndex]);
						}
					);
```
Or:
```suggestion
				auto [minEfficiencyIter, maxEfficiencyIter] = std::minmax_element(
					bestPal.begin(),
					bestPal.end(),
			        [&efficiency, &protoPalettes](
			        	ProtoPalAttrs const &lhs,
			            ProtoPalAttrs const &rhs
			        ) {
						return efficiency(protoPalettes[lhs.palIndex])
				    		< efficiency(protoPalettes[rhs.palIndex]);
					}
				);
```

Reply:
```cxx
				auto [minEfficiencyIter, maxEfficiencyIter] =
					std::minmax_element(
						bestPal.begin(),
						bestPal.end(),
						[&efficiency, &protoPalettes](
							ProtoPalAttrs const &lhs,
							ProtoPalAttrs const &rhs
						) {
							return efficiency(protoPalettes[lhs.palIndex])
								< efficiency(protoPalettes[rhs.palIndex]);
						});
```

Reply:
Yeah, that's what I was trying to post, but had to make a couple of edits and fix the github indentation.

Reply:
Glad that we agree on style without arguing, for once :D